chainservice/builder.go

```diff
@@ -0,0 +1,689 @@
+	"github.com/iotexproject/iotex-core/actpool"
+	"github.com/iotexproject/iotex-core/api"
+	"github.com/iotexproject/iotex-core/blockchain"
+	"github.com/iotexproject/iotex-core/blockchain/block"
+	"github.com/iotexproject/iotex-core/blockchain/blockdao"
+	"github.com/iotexproject/iotex-core/blockindex"
+	"github.com/iotexproject/iotex-core/blocksync"
+	"github.com/iotexproject/iotex-core/config"
+	"github.com/iotexproject/iotex-core/consensus"
+	"github.com/iotexproject/iotex-core/db"
+	"github.com/iotexproject/iotex-core/p2p"
+	"github.com/iotexproject/iotex-core/pkg/log"
+	"github.com/iotexproject/iotex-core/state/factory"
+	"github.com/iotexproject/iotex-election/committee"
+	"github.com/iotexproject/iotex-proto/golang/iotexrpc"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/proto"
+)
+
+// Builder is a builder to build chainservice
+type Builder struct {
+	cfg config.Config
+	cs  *ChainService
+}
+
+// NewBuilder creates a new chainservice builder
+func NewBuilder(cfg config.Config) *Builder {
+	builder := &Builder{cfg: cfg}
+	builder.createInstance()
+
+	return builder
+}
+
+// SetActionPool sets the action pool instance
+func (builder *Builder) SetActionPool(ap actpool.ActPool) *Builder {
+	builder.createInstance()
+	builder.cs.actpool = ap
+	return builder
+}
+
+// SetBlockchain sets the blockchain instance
+func (builder *Builder) SetBlockchain(bc blockchain.Blockchain) *Builder {
+	builder.createInstance()
+	builder.cs.chain = bc
+	return builder
+}
+
+// SetFactory sets the factory instance
+func (builder *Builder) SetFactory(f factory.Factory) *Builder {
+	builder.createInstance()
+	builder.cs.factory = f
+	return builder
+}
+
+// SetBlockDAO sets the blockdao instance
+func (builder *Builder) SetBlockDAO(bd blockdao.BlockDAO) *Builder {
+	builder.createInstance()
+	builder.cs.blockdao = bd
+	return builder
+}
+
+// SetP2PAgent sets the P2PAgent instance
+func (builder *Builder) SetP2PAgent(agent p2p.Agent) *Builder {
+	builder.createInstance()
+	builder.cs.p2pAgent = agent
+	return builder
+}
+
+// SetElectionCommittee sets the election committee instance
+func (builder *Builder) SetElectionCommittee(c committee.Committee) *Builder {
+	builder.createInstance()
+	builder.cs.electionCommittee = c
+	return builder
+}
+
+// SetBlockSync sets the block sync instance
+func (builder *Builder) SetBlockSync(bs blocksync.BlockSync) *Builder {
+	builder.createInstance()
+	builder.cs.blocksync = bs
+	return builder
+}
+
+// BuildForTest builds a chainservice for test purpose
+func (builder *Builder) BuildForTest() (*ChainService, error) {
+	builder.createInstance()
+	return builder.build(false, true)
+}
+
+// BuildForSubChain builds a chainservice for subchain
+func (builder *Builder) BuildForSubChain() (*ChainService, error) {
+	builder.createInstance()
+	return builder.build(true, false)
+}
+
+// Build builds a chainservice
+func (builder *Builder) Build() (*ChainService, error) {
+	builder.createInstance()
```

Comment:
similar to the first line of `build()`
```
builder.cs.registry = protocol.NewRegistry()
```
can add 1 line 
```
builder.cs = &ChainService{}
builder.cs.registry = protocol.NewRegistry()
```
then we don't need `createInstance()`

Reply:
Please double check the way `builder.cs` and `createInstance` are used.